specification/dartLangSpec.tex

```diff
@@ -10474,84 +10508,158 @@ \subsection{Assignment}
+  (\commentary{possibly implicitly induced by a variable}).
 
-\commentary{
-If $v$ is a final variable, a compile-time error has occurred and execution is unspecified.
-But a program with no compile-time errors may incur a dynamic type error.
-}
+  The expression $e$ is evaluated to an object $o$.
+  Then the setter $s$ is invoked
+  with its formal parameter bound to $o$.
+  If no error occurs, the value of $a$ is $o$.
 
-% add local functions per bug 23218
+  \commentary{%
+    $D$ cannot be a static setter in a class $C$,
+    because $a$ is then treated as
+    \code{$C$.\id{} = $e$},
+    which is specified elsewhere.%
+  }
+\item
+  \commentary{%
+    The case where the lexical lookup of \code{\id=}
+    yields a member signature cannot occur,
+    because that case is treated as
+    \code{\THIS.\id{} = $e$},
+    whose evaluation is specified elsewhere.
+  }
+\end{itemize}
 
 \LMHash{}%
-% TODO(eernst): $d$ defined ambiguously, re-check next sentence when fixing.
-If $d$ is the declaration of a library variable, top level getter or top level setter, the expression $e$ is evaluated to an object $o$.
-% TODO(eernst): $d$ defined ambiguously, re-check when fixing: Case $d$ is the getter and there is no setter.
-Then the setter \code{$v$=} is invoked with its formal parameter bound to $o$.
-The value of the assignment expression is $o$.
+\Case{\code{$p$.\id{} = $e$}}
+Consider an assignment $a$ of the form \code{$p$.\id{} = $e$},
+where $p$ is an import prefix and \id{} is an identifier.
 
 \LMHash{}%
-Otherwise, if $d$ is the declaration of a class variable, static getter or static setter in class $C$,
-then the assignment is equivalent to the assignment \code{$C$.$v$ = $e$}.
-
-\commentary{
-Otherwise, if $a$ occurs inside a top level or static function
-(be it function, method, getter, or setter) or variable initializer,
-a compile-time error has occurred.
-}
+A compile-time error occurs,
+unless $p$ has a member which is a setter $s$ named \code{id=}
+(\commentary{which may be implicitly induced by a variable declaration})
+such that the static type of $e$
+is assignable to the parameter type of $s$.
 
 \LMHash{}%
-%% TODO(eernst): We don't want to transform code and than complain, see if this
-%% can be reworded to rely on static checks such that it only happens when it
-%% works, or maybe that's already true.
-Otherwise, the assignment is equivalent to the assignment \code{\THIS{}.$v$ = $e$}.
+The static type of $a$ is the static type of $e$.
 
 \LMHash{}%
-% This error can occur due to implicit casts.
-It is a dynamic type error if the dynamic type of $o$
-is not a subtype of the actual type
-(\ref{actualTypeOfADeclaration})
-of $v$.
+\LMHash{}%
+Evaluation of an assignment $a$ of the form \code{$p$.\id{} = $e$}
+proceeds as follows:
+The expression $e$ is evaluated to an object $o$.
+Then the setter denoted by \code{$p$.\id} is invoked
+with its formal parameter bound to $o$.
+If no error occurs, the value of $a$ is $o$.
```

Comment:
> Then $a$ evaluates to the object $o$.

(same all similar places below, I'll stop saying it now).

Reply:
Done. There weren't any more places. ;-)